backend/lib/edgehog_web/gql_socket.ex

```diff
@@ -22,9 +22,82 @@ defmodule EdgehogWeb.GqlSocket do
   use Phoenix.Socket
   use Absinthe.Phoenix.Socket, schema: EdgehogWeb.Schema
 
-  def connect(_params, socket) do
-    {:ok, socket}
+  alias Ash.Error.Query.NotFound
+  alias Edgehog.Tenants
+
+  def connect(%{"Authorization" => token, "tenant" => tenant_slug}, socket, _connect_info) do
+    with {:ok, tenant} <- Tenants.fetch_tenant_by_slug(tenant_slug),
+         {:ok, claims} <- verify_jwt(token, tenant) do
+      socket =
+        socket
+        |> assign(:current_tenant, tenant)
+        |> assign(:claims, claims)
+        |> Absinthe.Phoenix.Socket.put_options(
+          context: %{
+            current_tenant: tenant,
+            claims: claims
+          }
+        )
+
+      {:ok, socket}
+    else
+      {:error, %Ash.Error.Invalid{errors: [%NotFound{} | _]}} ->
+        {:error, %{reason: "tenant_not_found"}}
+
+      {:error, %NotFound{}} ->
+        {:error, %{reason: "tenant_not_found"}}
+
+      {:error, :invalid_token} ->
+        {:error, %{reason: "invalid_token"}}
+
+      {:error, _reason} ->
+        {:error, %{reason: "unauthorized"}}
+    end
+  end
+
+  def connect(_params, _socket, _connect_info) do
+    {:error, %{reason: "unauthorized", message: "Missing params"}}
   end
 
   def id(_socket), do: nil
+
+  defp verify_jwt(token_header, tenant) do
+    token = String.replace_prefix(token_header, "Bearer ", "")
+
+    normalized_public_key = String.replace(tenant.public_key, "\\n", "\n")
+    jwk = JOSE.JWK.from_pem(normalized_public_key)
+    {_kty, jwk_map} = JOSE.JWK.to_map(jwk)
+
+    with {:ok, signer} <- signer_from_jwk_map(jwk_map),
+         {:ok, claims} <- Joken.verify(token, signer, []),
+         {:ok, claims} <- validate_jwt_claims(claims) do
+      {:ok, claims}
+    else
+      {:error, _reason} -> {:error, :invalid_token}
+    end
+  end
+
+  defp signer_from_jwk_map(%{"kty" => "RSA"} = jwk_map), do: {:ok, Joken.Signer.create("RS256", jwk_map)}
+
+  defp signer_from_jwk_map(%{"kty" => "EC"} = jwk_map), do: {:ok, Joken.Signer.create("ES256", jwk_map)}
+
+  defp signer_from_jwk_map(_), do: {:error, :unsupported_key_type}
+
+  defp validate_jwt_claims(%{"exp" => exp, "iat" => iat, "e_tga" => e_tga} = claims) when is_map(claims) do
+    with :ok <- check_e_tga_present(e_tga),
+         :ok <- check_jwt_expire(exp, iat) do
+      {:ok, claims}
+    end
+  end
```

Comment:
```suggestion
  defp validate_jwt_claims(%{"exp" => exp, , "e_tga" => e_tga} = claims) when is_map(claims) do
     case check_jwt_expire(exp) do
       :ok -> {:ok, claims}
       error -> error
      end
  end
  
  defp validate_jwt_claims(%{ "e_tga" => _e_tga} = claims) when is_map(claims) do
     {:ok, claims}
  end
  
   defp check_jwt_expire(exp) when is_integer(exp)  do
    if exp > Joken.current_time(), do: :ok, else: {:error, :token_expired}
  end

  defp check_jwt_expire(_exp), do: {:error, :invalid_claims}
```

As written, `validate_jwt_claims` will fail pattern matching when a token is generated without an `exp` claim, so that scenario should be handled explicitly. Additionally, comparing `exp` with `iat` would always succeed, since `iat` represents the token’s issue time and should always be earlier than `exp`.

Finally, a separate `check_e_tga_present` function doesn’t seem necessary here, as a successful pattern match already guarantees that the claim is present.